File: src/main/java/com/homeclimatecontrol/esphome2influxdb/Endpoint.java
package com.homeclimatecontrol.esphome2influxdb;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;

/**
 * TCP endpoint.
 */
public class Endpoint implements Verifiable {

    protected final Logger logger = LogManager.getLogger();

    private String name;
    public String host = "localhost";
    private int port;
    public String username;
    public String password;


    public int getPort() {
        return port;
    }

    public void setPort(int port) {
        this.port = port;
    }

    @Override
    public void verify() {

        var messages = new ArrayList<String>();

        if (host == null || host.isEmpty()) {
            messages.add("host can't be null or empty");
        }

        if (port <= 0) {
            messages.add("port can't be negative (" + port + " provided)");
        }

        if (!messages.isEmpty()) {

            var message = "Invalid state:";

            for (String s : messages) {
                message += "\n" + s;
            }

            throw new IllegalArgumentException(message);
        }
    }

    public String getName() {

        if (name != null) {
            return name;
        }

        return host + ":" + port;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public String toString() {

        var sb = new StringBuilder();

        sb.append("{");

        render(sb);

        sb.append("}");

        return sb.toString();
    }

    protected void render(StringBuilder sb) {

        sb.append("class=").append(getClass().getName());
        sb.append(",name=").append(getName());
        sb.append(",host:port=").append(host).append(":").append(port);

        if (username != null || password != null) {
            sb.append(",username:password").append(username).append(":").append(password);
        }
    }
}

File: src/main/kotlin/com/homeclimatecontrol/esphome2influxdb/k/Endpoint.kt
package com.homeclimatecontrol.esphome2influxdb.k

import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

open class Endpoint : Verifiable {
    protected val logger: Logger = LogManager.getLogger()

    private var name: String? = null
    var host: String? = "localhost"
    private var port = 0
    var username: String? = null
    var password: String? = null

    fun getPort(): Int {
        return port
    }

    fun setPort(port: Int) {
        this.port = port
    }

    override fun verify() {

        val messages: MutableList<String> = ArrayList()

        if (host == null || "" == host) {
            messages.add("host can't be null or empty")
        }

        if (port <= 0) {
            messages.add("port can't be negative ($port provided)")
        }

        if (messages.isNotEmpty()) {
            var message = "Invalid state:"
            for (s in messages) {
                message += """
                
                $s
                """.trimIndent()
            }
            throw IllegalArgumentException(message)
        }
    }

    fun getName(): String {
        return if (name != null) {
            name as String
        } else "$host:$port"
    }

    fun setName(name: String) {
        this.name = name
    }

    override fun toString(): String {
        return "{".plus(render()).plus("}")
    }

    protected open fun render() : String {

        val part1 = "class=".plus(javaClass.name)
            .plus(",name=").plus(getName())
            .plus(",host:port=$host:$port")

        if (username == null && password == null) {
            return part1;
        }

        return part1.plus(",username:password=$username:$password")
    }
}
